PWGJE/TableProducer/emcalCorrectionTask.cxx

```diff
@@ -174,238 +178,265 @@ struct EmcalCorrectionTask {
+              // TODO only consider tracks in current emcal/dcal acceptanc
+              if (!track.isGlobalTrack()) { // only global tracks
+                continue;
+              }
+              NTrack++;
+              if (hasPropagatedTracks) { // only temporarily while not every data has the tracks propagated to EMCal/PHOS
+                trackPhi.emplace_back(TVector2::Phi_0_2pi(track.trackPhiEmcal()));
+                trackEta.emplace_back(track.trackEtaEmcal());
+                hGlobalTrackEtaPhi->Fill(track.trackEtaEmcal(), TVector2::Phi_0_2pi(track.trackPhiEmcal()));
+              } else {
+                trackPhi.emplace_back(TVector2::Phi_0_2pi(track.phi()));
+                trackEta.emplace_back(track.eta());
+                hGlobalTrackEtaPhi->Fill(track.eta(), TVector2::Phi_0_2pi(track.phi()));
+              }
+              trackGlobalIndex.emplace_back(track.globalIndex());
             }
-            NTrack++;
-            if (hasPropagatedTracks) { // only temporarily while not every data has the tracks propagated to EMCal/PHOS
-              trackPhi.emplace_back(TVector2::Phi_0_2pi(track.trackPhiEmcal()));
-              trackEta.emplace_back(track.trackEtaEmcal());
-              hGlobalTrackEtaPhi->Fill(track.trackEtaEmcal(), TVector2::Phi_0_2pi(track.trackPhiEmcal()));
-            } else {
-              trackPhi.emplace_back(TVector2::Phi_0_2pi(track.phi()));
-              trackEta.emplace_back(track.eta());
-              hGlobalTrackEtaPhi->Fill(track.eta(), TVector2::Phi_0_2pi(track.phi()));
+            hGlobalTrackMult->Fill(NTrack);
+
+            std::vector<double> clusterPhi;
+            std::vector<double> clusterEta;
+
+            // TODO one loop that could in principle be combined with the other loop to improve performance
+            for (const auto& cluster : mAnalysisClusters) {
+              // Determine the cluster eta, phi, correcting for the vertex position.
+              auto pos = cluster.getGlobalPosition();
+              pos = pos - math_utils::Point3D<float>{vx, vy, vz};
+              // Normalize the vector and rescale by energy.
+              pos *= (cluster.E() / std::sqrt(pos.Mag2()));
+              clusterPhi.emplace_back(TVector2::Phi_0_2pi(pos.Phi()));
+              clusterEta.emplace_back(pos.Eta());
             }
-            trackGlobalIndex.emplace_back(track.globalIndex());
-          }
-          hGlobalTrackMult->Fill(NTrack);
-
-          std::vector<double> clusterPhi;
-          std::vector<double> clusterEta;
+            auto&& [clusterToTrackIndexMap, trackToClusterIndexMap] = JetUtilities::MatchClustersAndTracks(clusterPhi, clusterEta, trackPhi, trackEta, maxMatchingDistance, 10);
+            // we found a collision, put the clusters into the none ambiguous table
+            clusters.reserve(mAnalysisClusters.size());
+            int cellindex = -1;
+
+            unsigned int k = 0;
+            for (const auto& cluster : mAnalysisClusters) {
```

Comment:
Could we implement a JetUtilities::MatchClustersAndTracks function that only uses a single cluster or just always only provide a single cluster in the array so we could merge the two loops over the mAnalysisClusters? @fjonasALICE 

Other than that, this PR looks good from my side! 👍

Reply:
no I think this is not possible or at least i think we should not do it for single clusters. The system works by knowing positions of call clusters and track. Important is also (however not used yet) that the system works in both directions! It will not only return the closest tracks for each cluster but also all closest clusters for each track

Reply:
in any case, since number of clusters is not too large, i think the second loop does not hurt our performance

Reply:
Even if such an implementation would be possible and combine loops into 1 running MatchClustersAndTracks per cluster would come with a huge overhead, as the KD-Tree with tracks would have to be created for every cluster separately and would make the code extremely slow. I had similar performance losses in other tasks. Sometimes building maps and processing all in one go is performance-wise better even if it needs an extra loop for building the map.